### ISSUE TITLE
Bug/error with DQCP problem

### ISSUE THREAD

**Issue by nikkujensen** [cvxpy/cvxpy#2373]
**Describe the bug**
Canonicalization of PSD constraint returns error in DQCP problem.

**To Reproduce**
```
# import
import cvxpy as cp

# variable
A = cp.Variable((2,2),symmetric=True)
x = A[0,1]
y = A[1,1]

# assertions and constraints
x = cp.atoms.affine.wraps.nonneg_wrap(x)
y = cp.atoms.affine.wraps.nonneg_wrap(y)
constraints = [A>>0]

# function
f = x*y

# print function type
print("f(x,y) is ",f.sign,f.curvature)

# create the problem
problem = cp.Problem(cp.Minimize(-f),constraints)

# check if problem is DQCP
print("Is problem DQCP? ",problem.is_dqcp())

# solve
assert problem.is_dqcp()
problem.solve(qcp=True)
```

**Expected behavior**
I expect the problem to solve.

**Output**
```
psdErrorinDQCP.py"
f(x,y) is  NONNEGATIVE QUASICONCAVE
Is problem DQCP?  True
Traceback (most recent call last):
  File "/Users/njjensen/Documents/Nick's Drive/_My Drive/Work/Tokyo Tech Research/2022 MIC Project/2023-11 Three_N-Cornered Hat/psdErrorinDQCP.py", line 28, in <module>
    problem.solve(qcp=True)
  File "/Users/njjensen/miniconda3/envs/NCH/lib/python3.11/site-packages/cvxpy/problems/problem.py", line 503, in solve
    return solve_func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/njjensen/miniconda3/envs/NCH/lib/python3.11/site-packages/cvxpy/problems/problem.py", line 1068, in _solve
    chain.reduce(), solver=solver, verbose=verbose, **kwargs)
    ^^^^^^^^^^^^^^
  File "/Users/njjensen/miniconda3/envs/NCH/lib/python3.11/site-packages/cvxpy/reductions/reduction.py", line 94, in reduce
    problem, retrieval_data = self.apply(self.problem)
                              ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/njjensen/miniconda3/envs/NCH/lib/python3.11/site-packages/cvxpy/reductions/chain.py", line 76, in apply
    problem, inv = r.apply(problem)
                   ^^^^^^^^^^^^^^^^
  File "/Users/njjensen/miniconda3/envs/NCH/lib/python3.11/site-packages/cvxpy/reductions/dqcp2dcp/dqcp2dcp.py", line 101, in apply
    constraints += self._canonicalize_constraint(constr)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/njjensen/miniconda3/envs/NCH/lib/python3.11/site-packages/cvxpy/reductions/dqcp2dcp/dqcp2dcp.py", line 167, in _canonicalize_constraint
    rhs = constr.args[1]
          ~~~~~~~~~~~^^^
IndexError: list index out of range
```

**Version**
 - OS: macOS Sonoma 14.2.1
 - CVXPY Version: 1.4.2

**Additional context**
I can avoid this error by putting the PSD constraint with the definition of the matrix variable as `A = cp.Variable((2,2),PSD=True)` and the problem will run appropriately. **Note: this problem is just a minimum code example so there is no physical meaning**. Do not pay much attention to the physical meaning. Actually, as is the problem is unbounded.

**Comment by chofchof** [comment]
The `IndexError` occured at the method `_canonicalize_constraint(self, constr)` of the calss `Dqcp2Dcp`, because the length of the argument `constr` is just 1.

The type of `constr` is `<class 'cvxpy.constraints.psd.PSD'>` and the value is `var1 + Promote(-0.0, (2, 2)) >> 0`. Moreover, `constr.is_dcp()` is `True`. 

After changing the order of codes in `cvxpy/reductions/dqcp2dcp/dqcp2dcp.py` (line 166) as follows,
```
        if constr.is_dcp():
            canon_constr, aux_constr = self.canonicalize_tree(constr)
            return [canon_constr] + aux_constr

        lhs = constr.args[0]
        rhs = constr.args[1]

        if isinstance(constr, Inequality):
```
I could get the following result (with `verbose=True`) with `cvxpy.error.SolverError: Max iters hit during bisection.` 
```
f(x,y) is  NONNEGATIVE QUASICONCAVE
Is problem DQCP?  True
===============================================================================
                                     CVXPY
                                     v1.4.3
===============================================================================
(CVXPY) Apr 22 03:50:15 PM: Your problem has 4 variables, 1 constraints, and 0 parameters.
(CVXPY) Apr 22 03:50:15 PM: It is compliant with the following grammars: DQCP
(CVXPY) Apr 22 03:50:15 PM: (If you need to solve this problem multiple times, but with different data, consider using parameters.)
(CVXPY) Apr 22 03:50:15 PM: CVXPY will first compile your problem; then, it will invoke a numerical solver to obtain a solution.
(CVXPY) Apr 22 03:50:15 PM: Your problem is compiled with the CPP canonicalization backend.
(CVXPY) Apr 22 03:50:15 PM: Reducing DQCP problem to a one-parameter family of DCP problems, for bisection.

********************************************************************************
Preparing to bisect problem

minimize 0.0
subject to var1 + Promote(-0.0, (2, 2)) >> 0
           -param15 @ power(nonneg_wrap(var1[1, 1]), -1.0) <= nonneg_wrap(var1[0, 1])

Finding interval for bisection ...
initial lower bound: -70368744177664.000000
initial upper bound: -35184372088832.000000

(iteration 0) lower bound: -70368744177664.000000
(iteration 0) upper bound: -35184372088832.000000
(iteration 0) query point: -52776558133248.000000
(iteration 0) query was infeasible.

(iteration 5) lower bound: -45079976738816.000000
(iteration 5) upper bound: -43980465111040.000000
(iteration 5) query point: -44530220924928.000000
(iteration 5) query was infeasible.

...(SKIP)...

(iteration 95) lower bound: -44076302993073.890625
(iteration 95) upper bound: -44076302993073.882812
(iteration 95) query point: -44076302993073.890625
(iteration 95) query was infeasible.

Traceback (most recent call last):
  File "test_2373.py", line 28, in <module>
    problem.solve(qcp=True, verbose=True)
  File "/...(SKIP).../lib/python3.12/site-packages/cvxpy/problems/problem.py", line 503, in solve
    return solve_func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/...(SKIP).../lib/python3.12/site-packages/cvxpy/problems/problem.py", line 1067, in _solve
    soln = bisection.bisect(
           ^^^^^^^^^^^^^^^^^
  File "/...(SKIP).../lib/python3.12/site-packages/cvxpy/reductions/solvers/bisection.py", line 189, in bisect
    soln, low, high = _bisect(
                      ^^^^^^^^
  File "/...(SKIP).../lib/python3.12/site-packages/cvxpy/reductions/solvers/bisection.py", line 134, in _bisect
    raise error.SolverError("Max iters hit during bisection.")
cvxpy.error.SolverError: Max iters hit during bisection.
```

**Comment by SteveDiamond** [comment]
Fixed by #2424
